# Домашнее задание 2

Используя датасет [STUDENT ALCOHOL CONSUMPTION](https://archive.ics.uci.edu/ml/datasets/STUDENT+ALCOHOL+CONSUMPTION#) и статью "USING	DATA	MINING	TO	PREDICT	SECONDARY SCHOOL	STUDENT	ALCOHOL	CONSUMPTION" выполните следующие задания:

In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

### Предобработка (Max 2)
- Приведите описание признаков датасета на русском языке с указанием типа данных (0.3 балла)
- Предобработка данных (1.7)
    - Проверьте наличие пропусков. В случае наличия пропусков заполните их медианными значениями (0.4)
    - Подсчитайте количество людей, у которых матери с образованием "0" (0.4)
    - Преобразуйте все номинальные признаки в несколько признаков с бинарным значение (0.4)
    - Используя объект [SelectKBest](http://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectKBest.html#sklearn.feature_selection.SelectKBest) с параметром [f_regression](http://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_regression.html#sklearn.feature_selection.f_regression)  выберите два признака, которые наиболее сильно связаны с признаком G3 (0.5)

In [4]:
descr = open('data/student.txt')
for i in descr:
    print(i)

# Attributes for both student-mat.csv (Math course) and student-por.csv (Portuguese language course) datasets:

1 school - student's school (binary: "GP" - Gabriel Pereira or "MS" - Mousinho da Silveira)

2 sex - student's sex (binary: "F" - female or "M" - male)

3 age - student's age (numeric: from 15 to 22)

4 address - student's home address type (binary: "U" - urban or "R" - rural)

5 famsize - family size (binary: "LE3" - less or equal to 3 or "GT3" - greater than 3)

6 Pstatus - parent's cohabitation status (binary: "T" - living together or "A" - apart)

7 Medu - mother's education (numeric: 0 - none,  1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education)

8 Fedu - father's education (numeric: 0 - none,  1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education)

9 Mjob - mother's job (nominal: "teacher", "health" care related, civil "services" (e.g. administrative or police), "at_

In [170]:
data_mat = pd.read_csv('data/student-mat.csv', sep=';')
data_por = pd.read_csv('data/student-por.csv', sep=';')

In [151]:
data_por.head(3)

age  Medu  Fedu  traveltime  studytime  failures  famrel  freetime  goout  \
0   18     4     4           2          2         0       4         3      4   
1   17     1     1           1          2         0       5         3      3   
2   15     1     1           1          2         0       4         3      2   

   Dalc      ...       activities_no  activities_yes  nursery_no  nursery_yes  \
0     1      ...                   1               0           0            1   
1     1      ...                   1               0           1            0   
2     2      ...                   1               0           0            1   

   higher_no  higher_yes  internet_no  internet_yes  romantic_no  romantic_yes  
0          0           1            1             0            1             0  
1          0           1            0             1            1             0  
2          0           1            0             1            1             0  

[3 rows x 59 columns]

In [152]:
data_mat.head(3)

age  Medu  Fedu  traveltime  studytime  failures  famrel  freetime  goout  \
0   18     4     4           2          2         0       4         3      4   
1   17     1     1           1          2         0       5         3      3   
2   15     1     1           1          2         3       4         3      2   

   Dalc      ...       activities_no  activities_yes  nursery_no  nursery_yes  \
0     1      ...                   1               0           0            1   
1     1      ...                   1               0           1            0   
2     2      ...                   1               0           0            1   

   higher_no  higher_yes  internet_no  internet_yes  romantic_no  romantic_yes  
0          0           1            1             0            1             0  
1          0           1            0             1            1             0  
2          0           1            0             1            1             0  

[3 rows x 59 columns]

In [116]:
data_por.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  649.000000  649.000000  649.000000  649.000000  649.000000  649.000000   
mean    16.744222    2.514638    2.306626    1.568567    1.930663    0.221880   
std      1.218138    1.134552    1.099931    0.748660    0.829510    0.593235   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    1.000000    1.000000    1.000000    0.000000   
50%     17.000000    2.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  649.000000  649.000000  649.000000  649.000000  649.000000  649.000000   
mean     3.930663    3.180277    3.184900    1.502311    2.280431    3.536210   
std      0.955717    1.051093    1.175766    0.924834    1.284380    1.446259   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    2.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3  
count  649.000000  649.000000  649.000000  649.000000  
mean     3.659476   11.399076   11.570108   11.906009  
std      4.640759    2.745265    2.913639    3.230656  
min      0.000000    0.000000    0.000000    0.000000  
25%      0.000000   10.000000   10.000000   10.000000  
50%      2.000000   11.000000   11.000000   12.000000  
75%      6.000000   13.000000   13.000000   14.000000  
max     32.000000   19.000000   19.000000   19.000000

In [118]:
data_mat.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.998862    1.113278    0.890741    1.287897    1.390303   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3  
count  395.000000  395.000000  395.000000  395.000000  
mean     5.708861   10.908861   10.713924   10.415190  
std      8.003096    3.319195    3.761505    4.581443  
min      0.000000    3.000000    0.000000    0.000000  
25%      0.000000    8.000000    9.000000    8.000000  
50%      4.000000   11.000000   11.000000   11.000000  
75%      8.000000   13.000000   13.000000   14.000000  
max     75.000000   19.000000   19.000000   20.000000

In [156]:
data_mat[data_mat.isnull().any(axis=1)]

Empty DataFrame
Columns: [age, Medu, Fedu, traveltime, studytime, failures, famrel, freetime, goout, Dalc, Walc, health, absences, G1, G2, G3, school_GP, school_MS, sex_F, sex_M, address_R, address_U, famsize_GT3, famsize_LE3, Pstatus_A, Pstatus_T, Mjob_at_home, Mjob_health, Mjob_other, Mjob_services, Mjob_teacher, Fjob_at_home, Fjob_health, Fjob_other, Fjob_services, Fjob_teacher, reason_course, reason_home, reason_other, reason_reputation, guardian_father, guardian_mother, guardian_other, schoolsup_no, schoolsup_yes, famsup_no, famsup_yes, paid_no, paid_yes, activities_no, activities_yes, nursery_no, nursery_yes, higher_no, higher_yes, internet_no, internet_yes, romantic_no, romantic_yes]
Index: []

[0 rows x 59 columns]

In [157]:
data_por[data_por.isnull().any(axis=1)]

Empty DataFrame
Columns: [age, Medu, Fedu, traveltime, studytime, failures, famrel, freetime, goout, Dalc, Walc, health, absences, G1, G2, G3, school_GP, school_MS, sex_F, sex_M, address_R, address_U, famsize_GT3, famsize_LE3, Pstatus_A, Pstatus_T, Mjob_at_home, Mjob_health, Mjob_other, Mjob_services, Mjob_teacher, Fjob_at_home, Fjob_health, Fjob_other, Fjob_services, Fjob_teacher, reason_course, reason_home, reason_other, reason_reputation, guardian_father, guardian_mother, guardian_other, schoolsup_no, schoolsup_yes, famsup_no, famsup_yes, paid_no, paid_yes, activities_no, activities_yes, nursery_no, nursery_yes, higher_no, higher_yes, internet_no, internet_yes, romantic_no, romantic_yes]
Index: []

[0 rows x 59 columns]

### No NaNs detected!

In [120]:
data_mat[data_mat['Medu'] == 0]

school sex  age address famsize Pstatus  Medu  Fedu     Mjob     Fjob ...  \
127     GP   F   19       U     GT3       T     0     1  at_home    other ...   
249     GP   M   16       U     GT3       T     0     2    other    other ...   
324     GP   F   17       U     LE3       T     0     2  at_home  at_home ...   

    famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
127      3        4      2     1     1      5        2   7   8   9  
249      4        3      2     2     4      5        0  13  15  15  
324      3        3      3     2     3      2        0  16  15  15  

[3 rows x 33 columns]

In [121]:
data_por[data_por['Medu'] == 0]

school sex  age address famsize Pstatus  Medu  Fedu     Mjob     Fjob ...  \
127     GP   F   19       U     GT3       T     0     1  at_home    other ...   
281     GP   M   16       U     GT3       T     0     2    other    other ...   
375     GP   F   17       U     LE3       T     0     2  at_home  at_home ...   
437     MS   F   16       R     GT3       T     0     2    other    other ...   
493     MS   F   17       U     GT3       T     0     1    other  at_home ...   
584     MS   F   17       R     GT3       T     0     0  at_home    other ...   

    famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
127      3        4      2     1     1      5        0   9  10  11  
281      4        3      2     2     4      5        0  11  12  11  
375      3        3      3     2     3      2        0  14  14  15  
437      3        2      3     1     2      2        0  12  11  12  
493      2        4      4     3     5      5        5   9   9  10  
584      4        4      3     1     1      5        0  10  11  11  

[6 rows x 33 columns]

Binary and Categorial features to transform:
* school - student's school (binary: "GP" - Gabriel Pereira or "MS" - Mousinho da Silveira)<p></p>

* sex - student's sex (binary: "F" - female or "M" - male)<p></p>

* address - student's home address type (binary: "U" - urban or "R" - rural)<p></p>

* famsize - family size (binary: "LE3" - less or equal to 3 or "GT3" - greater than 3)<p></p>

* Pstatus - parent's cohabitation status (binary: "T" - living together or "A" - apart)<p></p>

* Mjob - mother's job (nominal: "teacher", "health" care related, civil "services" (e.g. administrative or police), "at_home" or "other")<p></p>

* Fjob - father's job (nominal: "teacher", "health" care related, civil "services" (e.g. administrative or police), "at_home" or "other")<p></p>

* reason - reason to choose this school (nominal: close to "home", school "reputation", "course" preference or "other")<p></p>

* guardian - student's guardian (nominal: "mother", "father" or "other")<p></p>

* schoolsup - extra educational support (binary: yes or no)<p></p>

* famsup - family educational support (binary: yes or no)<p></p>

* paid - extra paid classes within the course subject (Math or Portuguese) (binary: yes or no)<p></p>

* activities - extra-curricular activities (binary: yes or no)<p></p>

* nursery - attended nursery school (binary: yes or no)<p></p>

* higher - wants to take higher education (binary: yes or no)<p></p>

* internet - Internet access at home (binary: yes or no)<p></p>

* romantic - with a romantic relationship (binary: yes or no)<p></p>


In [168]:
def convert_to_binary(d, strg):
    d[strg] = d[strg].astype('category')
    d[strg] = d[strg].cat.codes
tmp = ['school', 'sex', 'address', 'famsize', 'Pstatus', 'schoolsup', 'famsup', 'paid',
      'activities', 'nursery', 'higher', 'internet', 'romantic']

In [171]:
[convert_to_binary(data_mat, i) for i in tmp]
[convert_to_binary(data_por, i) for i in tmp]
data_mat = pd.get_dummies(data_mat, columns=['Mjob', 'Fjob', 'reason', 'guardian'])
data_por = pd.get_dummies(data_por, columns=['Mjob', 'Fjob', 'reason', 'guardian'])

In [173]:
data_mat.dtypes

school                int8
sex                   int8
age                  int64
address               int8
famsize               int8
Pstatus               int8
Medu                 int64
Fedu                 int64
traveltime           int64
studytime            int64
failures             int64
schoolsup             int8
famsup                int8
paid                  int8
activities            int8
nursery               int8
higher                int8
internet              int8
romantic              int8
famrel               int64
freetime             int64
goout                int64
Dalc                 int64
Walc                 int64
health               int64
absences             int64
G1                   int64
G2                   int64
G3                   int64
Mjob_at_home         uint8
Mjob_health          uint8
Mjob_other           uint8
Mjob_services        uint8
Mjob_teacher         uint8
Fjob_at_home         uint8
Fjob_health          uint8
Fjob_other           uint8
F

In [174]:
data_por.dtypes

school                int8
sex                   int8
age                  int64
address               int8
famsize               int8
Pstatus               int8
Medu                 int64
Fedu                 int64
traveltime           int64
studytime            int64
failures             int64
schoolsup             int8
famsup                int8
paid                  int8
activities            int8
nursery               int8
higher                int8
internet              int8
romantic              int8
famrel               int64
freetime             int64
goout                int64
Dalc                 int64
Walc                 int64
health               int64
absences             int64
G1                   int64
G2                   int64
G3                   int64
Mjob_at_home         uint8
Mjob_health          uint8
Mjob_other           uint8
Mjob_services        uint8
Mjob_teacher         uint8
Fjob_at_home         uint8
Fjob_health          uint8
Fjob_other           uint8
F

from sklearn.feature_selection import SelectKBest, f_regression
data_best_por = SelectKBest(f_regression, k=2).fit_transform(data_por, data_por['G3'])
data_best_mat = SelectKBest(f_regression, k=2).fit_transform(data_mat, data_mat['G3'])

In [240]:
from sklearn.feature_selection import SelectKBest, f_regression
data = pd.concat([data_mat, data_por], ignore_index=True)
data_best = SelectKBest(f_regression, k=2).fit_transform(data.drop('G3', axis=1), data['G3'])

In [242]:
print(data_best.shape)
data_best[0:3]

(1044, 2)


array([[5, 6],
       [5, 5],
       [7, 8]])

In [265]:
for i in data.columns:
    print(i, '-', data[i].head(3).values)

school - [0 0 0]
sex - [0 0 0]
age - [18 17 15]
address - [1 1 1]
famsize - [0 0 1]
Pstatus - [0 1 1]
Medu - [4 1 1]
Fedu - [4 1 1]
traveltime - [2 1 1]
studytime - [2 2 2]
failures - [0 0 3]
schoolsup - [1 0 1]
famsup - [0 1 0]
paid - [0 0 1]
activities - [0 0 0]
nursery - [1 0 1]
higher - [1 1 1]
internet - [0 1 1]
romantic - [0 0 0]
famrel - [4 5 4]
freetime - [3 3 3]
goout - [4 3 2]
Dalc - [1 1 2]
Walc - [1 1 3]
health - [3 3 3]
absences - [ 6  4 10]
G1 - [5 5 7]
G2 - [6 5 8]
G3 - [ 6  6 10]
Mjob_at_home - [1 1 1]
Mjob_health - [0 0 0]
Mjob_other - [0 0 0]
Mjob_services - [0 0 0]
Mjob_teacher - [0 0 0]
Fjob_at_home - [0 0 0]
Fjob_health - [0 0 0]
Fjob_other - [0 1 1]
Fjob_services - [0 0 0]
Fjob_teacher - [1 0 0]
reason_course - [1 1 0]
reason_home - [0 0 0]
reason_other - [0 0 1]
reason_reputation - [0 0 0]
guardian_father - [0 1 0]
guardian_mother - [1 0 1]
guardian_other - [0 0 0]


### SelectKBest выбрал G1 и G2 как лучшие признаки для предсказания G3;
*что в принципе не более чем очевидно*

### Регрессия (Max 4)
Выберите один признак, который наиболее влияет на признак G3 и используя кросс-валидацию (cross-validation) выполните следующее:

- На обучающей выборке обучите объект [RidgeСV](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RidgeCV.html#sklearn.linear_model.RidgeCV) для различных alpha (0.2)
- Выберите alpha при котором RidgeCV лучше всего предсказывает значения (0.1)
- Подсчитайте метрику mean absolute error, mean squared error, median absolute error,r2 score(0.2)
- Max (0.5)
- ------------------------------------
- Создайте объекты [SGDRegressor](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDRegressor.html#sklearn.linear_model.SGDRegressor), используя все возможных варианты аргумента penalty (0.2)
- Обучите все полученных объекты на обучающей выборке (0.1)
- Подсчитайте метрику r2 score для всех объектов (0.2)
- Max (0.5)
- ------------------------------------
- Создайте объекты с параметрами max_depth = 1,3, 10 [GradientBoostingRegressor](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingRegressor.html#sklearn.ensemble.GradientBoostingRegressor) и для каждого объекта изменяйте количество n_estimators  от 100 до 1000 с шагом 100 (0.2)
- Подсчитайте метрику r2 score для всех объектов (0.1)
- Визуализируйте зависимость R2 oт n_estimators на графике (0.2)
- Max (0.5)
- -----------------------------------
- Используя все доступные регрессоры в библиотеке sklearn, найдите самый лучший регрессор. (2.5)

### Классификация (Max 4)
- В статье описано использование Random Forest для предсказания важности факторов, влияющих на потребление алкоголя. Повторите эксперимент с использование объекта [RandomForest](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier) и найдите наилучшие параметры. Укажите accuracy для наилучших параметров (0.5)
- Обучите объект [GradientBoostingClassifier](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingClassifier.html#sklearn.ensemble.GradientBoostingClassifier)  и найдите для него наилучшие параметры. Укажите accuracy для наилучших параметров (0.5)
- Обучите объект [AdaBoostClassifier](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.AdaBoostClassifier.html)  и найдите для него наилучшие параметры. Укажите accuracy для наилучших параметров (0.5)
- Найдите наилучший классификатор. Укажите accuracy для наилучших параметров (2.5)

# Замечания

- В работе следует использовать библиотеку scikit-learn версии 0.18 и scipy версии 0.18.1